src/java/org/apache/cassandra/index/sai/disk/v1/trie/TrieTermsDictionaryReader.java

```diff
@@ -0,0 +1,284 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.cassandra.index.sai.disk.v1.trie;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.Arrays;
+import java.util.Iterator;
+import javax.annotation.concurrent.NotThreadSafe;
+
+import com.google.common.collect.AbstractIterator;
+
+import org.apache.cassandra.io.tries.SerializationNode;
+import org.apache.cassandra.io.tries.TrieNode;
+import org.apache.cassandra.io.tries.TrieSerializer;
+import org.apache.cassandra.io.tries.Walker;
+import org.apache.cassandra.io.util.DataOutputPlus;
+import org.apache.cassandra.io.util.Rebufferer;
+import org.apache.cassandra.utils.Pair;
+import org.apache.cassandra.utils.SizedInts;
+import org.apache.cassandra.utils.bytecomparable.ByteComparable;
+import org.apache.cassandra.utils.bytecomparable.ByteSource;
+import org.apache.lucene.util.ArrayUtil;
+
+/**
+ * Page-aware random access reader for a trie terms dictionary written by {@link TrieTermsDictionaryWriter}.
+ */
+@NotThreadSafe
+public class TrieTermsDictionaryReader extends Walker<TrieTermsDictionaryReader>
+{
+    public static final long NOT_FOUND = -1;
+
+    public TrieTermsDictionaryReader(Rebufferer rebufferer, long root)
+    {
+        super(rebufferer, root);
+    }
+
+    public static final TrieSerializer<Long, DataOutputPlus> trieSerializer = new TrieSerializer<Long, DataOutputPlus>()
+    {
+        @Override
+        public int sizeofNode(SerializationNode<Long> node, long nodePosition)
+        {
+            return TrieNode.typeFor(node, nodePosition).sizeofNode(node) + sizeof(node.payload());
+        }
+
+        @Override
+        public void write(DataOutputPlus dest, SerializationNode<Long> node, long nodePosition) throws IOException
+        {
+            final TrieNode type = TrieNode.typeFor(node, nodePosition);
+            final Long payload = node.payload();
+            if (payload != null)
+            {
+                final int payloadBits = SizedInts.nonZeroSize(payload);
+                type.serialize(dest, node, payloadBits, nodePosition);
+                SizedInts.write(dest, payload, payloadBits);
+            }
+            else
+            {
+                type.serialize(dest, node, 0, nodePosition);
+            }
+        }
+
+        private int sizeof(Long payload)
+        {
+            if (payload != null)
+            {
+                return SizedInts.nonZeroSize(payload);
+            }
+            return 0;
+        }
+    };
+
+    public long exactMatch(ByteComparable key)
+    {
+        int b = follow(key);
+        if (b != ByteSource.END_OF_STREAM)
+        {
+            return NOT_FOUND;
+        }
+        return getCurrentPayload();
```

Comment:
nit: `return b == ByteSource.END_OF_STREAM ? getCurrentPayload() : NOT_FOUND;`

I had to do a quick double-take there, realizing that `follow()` returns the first unmatched byte, which is EOS if we have a successful match. Usually `END_OF_X` will connote that we ran out of search space :D

Reply:
Yes, since this is an exact match then we always expect the result of `follow` to be EOS if it was successful. A bit counter-intuitive but I'll add a comment to clarify this as well.